### PR TITLE
WIP: Introduce concept of main and user repositories

### PR DESCRIPTION
**Pull request by lionel-**
Part of #781.

I'm currently creating a new package so I thought I'd go ahead and try to make usethis compatible with the origin + upstream workflow.

One idea explored in this tentative PR is to introduce the concepts of main and user repos.

- The main repo is where official development takes place. By default it is `upstream || origin`.
- The user repo is where local development takes place. By default it is `origin || upstream`.

For instance, Travis activation should use the main repo by default (it is occasionally useful to activate it in the user repo but I'm not sure how to accomodate this with the current UI). On the other hand, checking if Github is properly configured should be done with the user repo. I expect most usethis functions should interact with the main repo rather than the user repo.

### CODE REVIEW

**Comment by jennybc** [comment]
Just recording some of my own thoughts to park here.

I see three different scenarios:

  * Maintainer 1:
    - User is package maintainer or has write or admin permissions.
    - `origin` is the main repo. If there are other remotes, they are contributor remotes. No `upstream`.
    - Jenny and Hadley usually work this way.
  * Maintainer 2:
    - User is package maintainer or has write or admin permissions.
    - `origin` is a personal fork. `upstream` in the main or source repo.
    - Lionel usually works this way.
  * Contributor:
    - User does not have write or admin permissions.
    - `origin` is a personal fork. `upstream` in the main or source repo.
    - Typical contributor.

Basically our mentality so far has accounted for Maintainer 1 and Contributor. We didn't really plan for Maintainer 2.

**Comment by hadley** [comment]
If we go down this route, I think it probably makes sense to eliminate `github_upstream()` (and friends) and only refer to `github_main()` or `github_user()`.

**Comment by hadley** [comment]
Or maybe `github_upstream()` should always fall back to origin if upstream doesn't exist?

@lionel- when would you expect an upstream remote without an origin remote?

**Comment by lionel-** [comment]
If I have created the github repo instead of cloning it, I would call it "upstream" and not "origin". I would only call clones "origin". What do you think about these conventions @jimhester?

**Comment by jimhester** [comment]
The most widely used convention and what git defaults to is.

1. The place you push to is `origin`
2. The place you pull from is `upstream`
3. If there is only one of the above it is `origin`

I think doing anything else in usethis will lead people down a bad path.

**Comment by lionel-** [comment]
That makes sense. This suggests these rules for the main and user repos:

- Main repo: `upstream || origin`
- User repo: `origin`

Note that this is compatible with the rules suggested above:

- Main repo: `upstream || origin`
- User repo: `origin || upstream`

**Comment by hadley** [comment]
I think maybe using "push" and "pull" as name might help us remember. (But the main change we need to make is for upstream to always fall back to origin)

**Comment by jennybc** [comment]
> I think doing anything else or using other terms in usethis will lead people down a bad path.

I agree that adopting any sort of nonstandard convention, no matter how sane, will ultimately cause more pain than good. I tried to have my own convention around remote names for a while and eventually just gave up.

> I think maybe using "push" and "pull" as name might help us remember. (But the main change we need to make is for upstream to always fall back to origin)

One wrinkle is also that the push/pull setup will vary by branch. In a fork, `master` will pull from `upstream` (at least it will if it was created by `create_from_github()`. But `master` will push to `origin` (and this push should really never happen because no one should ever work in `master` of a fork). Non-`master` branches will pull from **and** push to `origin`.

**Comment by jennybc** [comment]
This is now handled by the changes that came as part of the huge gert PR #1144

**Comment by lionel-** [comment]
Thanks Jenny!
